### PR TITLE
Update SyncManager to handle checkpoint and block submitted events #97

### PR DESCRIPTION
**Pull request by syuhei176**
## Description

- [x] handle checkpoint event

### CODE REVIEW

File: client/src/sync/sync_manager.rs [hunk 1]
```diff
@@ -1,44 +1,105 @@
...
 where
     T: SyncDbTrait,
 {
-    /// Callback which is called when new block is submitted
-    pub fn sync(&self) -> Vec<StateQueryResult> {
+    pub fn initialize(&mut self, kvs: &'a Bucket<Vec<u8>>) {
+        let commitment_contracts = self.sync_db.get_commitment_contracts();
+        let deposit_contracts = self.get_deposit_contracts(commitment_contracts);
+
+        let abi: Vec<Event> = vec![Event {
```
Comment (syuhei176) [inline client/src/sync/sync_manager.rs]: I didn't get how I can specify abi of checkpoint event.

Comment (syuhei176) [reply client/src/sync/sync_manager.rs]: Solved!

File: client/src/sync/sync_manager.rs [hunk 2]
```diff
@@ -1,44 +1,116 @@
...
+            name: "LogCheckpoint".to_owned(),
+            inputs: vec![],
+            anonymous: false,
+        }];
+
+        for d in deposit_contracts {
+            let db = EventDbImpl::from(kvs.bucket(&d.as_bytes().into()));
+            let mut watcher = EventWatcher::new(&self.mainchain_endpoint, d, abi.clone(), db);
```
Comment (syuhei176) [inline client/src/sync/sync_manager.rs]:
```
   = note: expected type `std::vec::Vec<ethabi::event::Event>`
              found type `std::vec::Vec<state::state_db::ethabi::Event>`
note: Perhaps two different versions of crate `ethabi` are being used?
```

Comment (tkmct) [reply client/src/sync/sync_manager.rs]: Yes, that's what I wrote in [here](https://github.com/cryptoeconomicslab/plasma-rust-framework/pull/101#issuecomment-507110012)
web3 depends on ethabi 7.0.0

Comment (syuhei176) [reply client/src/sync/sync_manager.rs]: I wanna use patch.
https://doc.rust-lang.org/cargo/reference/manifest.html#the-patch-section

**Comment by tkmct** [comment]
Implement Decoder, Encoder for Checkpoint

File: client/src/sync/sync_manager.rs [hunk 3]
```diff
@@ -1,44 +1,117 @@
...
+                )
+                .ok()
+                .unwrap();
+                let state_update_tuple = decoded[0].clone().to_tuple();
+                let _range_tuple = decoded[1].clone().to_tuple();
+                let _state_update = StateUpdate::from_tuple(&state_update_tuple.unwrap())
+                    .ok()
+                    .unwrap();
+                /*
+                state_manager.deposit(
```
Comment (syuhei176) [inline client/src/sync/sync_manager.rs]: `K` cannot be sent between threads safely.

File: client/src/sync/sync_manager.rs (hunk 2 above)
Comment (syuhei176) [reply client/src/sync/sync_manager.rs]: Fixed by using the patched version.

```
web3 = { git = "https://github.com/cryptoeconomicslab/rust-web3", branch = "ethabi-tuple" }
```

File: client/src/sync/sync_manager.rs (hunk 1 above)
Comment (syuhei176) [reply client/src/sync/sync_manager.rs]: We will add checkpoint data structure.
https://github.com/cryptoeconomicslab/plasma-rust-framework/issues/106

File: client/src/sync/sync_manager.rs (hunk 3 above)
Comment (tkmct) [reply client/src/sync/sync_manager.rs]: I'll look at it

File: client/src/sync/sync_manager.rs [hunk 4]
```diff
@@ -166,4 +275,18 @@ mod tests {
             .is_ok());
     }
 
+    #[test]
+    fn test_initialized() {
```
Comment (syuhei176) [inline client/src/sync/sync_manager.rs]: Added initialize sync manager and polling test.

**Comment by syuhei176** [comment]
### Setting test environment

```
ganache-cli --mnemonic 'candy maple cake sugar pudding cream honey rich smooth crumble sweet treat'
```

```
./node_modules/.bin/truffle migrate --network local
node ./script/deposit.js 
```

* https://github.com/cryptoeconomicslab/plasma-predicates/blob/debug/script/deposit.js

**Pull request closed without merging**